CsvInput parses data in parallel

tinwelint opened the pull request:
takes advantage of being able to determine where any data row starts
and another ends if multi-line fields are disallowed. In this case
chunks of csv data can be read and handed off to one or more threads
parsing the data, to finally end up merged back in the same order as
they were read from the csv source.
From the outside there's no difference, it still provides `InputIterator`
for nodes and relationships.

changelog: [3.0, import-tool] Improves the performance of CSV input data reading by parsing data in parallel if multi-line fields are disallowed.

tinwelint committed:
Ability to clone CSV header data structure

since each contain Extractor which is state-full. In a parallel processing
scenario those must be cloned to each processing thread.

tinwelint committed:
Some class extractions and removal of unused code

tinwelint committed:
Pluggable reading of chunks in BufferedCharSeeker

so that other types of scenarios can be implemented,
not just the single threaded scenario.

tinwelint committed:
CsvInput parses data in parallel

takes advantage of being able to determine where any data row starts
and another ends if multi-line fields are disallowed. In this case
chunks of csv data can be read and handed off to one or more threads
parsing the data, to finally end up merged back in the same order as
they were read from the csv source.
From the outside there's no difference, it still provides `InputIterator`
for nodes and relationships.

tinwelint committed:
Processing queue limited to number of active processors

there's no need to queue more items than that, doing so will most likely
only cause higher memory usage to no gain.

tinwelint committed:
Added CsvOutput

which serializes any `Input` to CSV files.

tinwelint commented:
I don't get it - `CsvInputTest.shouldCloseDataIteratorsInTheEnd` fails with:

```
org.mockito.exceptions.base.MockitoException: 
Mockito cannot mock this class: class org.neo4j.csv.reader.Readables$3
Mockito can only mock visible & non-final classes.
If you're not sure why you're getting this error, please report to the mailing list.
    at java.lang.ClassLoader.defineClass1(Native Method)
    at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
    at sun.reflect.GeneratedMethodAccessor22.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.mockito.cglib.core.ReflectUtils.defineClass(ReflectUtils.java:385)
    at org.mockito.cglib.core.AbstractClassGenerator.create(AbstractClassGenerator.java:220)
    at org.mockito.cglib.proxy.Enhancer.createHelper(Enhancer.java:378)
    at org.mockito.cglib.proxy.Enhancer.createClass(Enhancer.java:318)
    at org.mockito.internal.creation.cglib.ClassImposterizer.createProxyClass(ClassImposterizer.java:123)
    at org.mockito.internal.creation.cglib.ClassImposterizer.imposterise(ClassImposterizer.java:57)
    at org.mockito.internal.creation.cglib.ClassImposterizer.imposterise(ClassImposterizer.java:49)
    at org.mockito.internal.creation.cglib.CglibMockMaker.createMock(CglibMockMaker.java:24)
    at org.mockito.internal.util.MockUtil.createMock(MockUtil.java:33)
    at org.mockito.internal.MockitoCore.mock(MockitoCore.java:59)
    at org.mockito.Mockito.spy(Mockito.java:1368)
    at org.neo4j.unsafe.impl.batchimport.input.csv.CsvInputTest.shouldCloseDataIteratorsInTheEnd(CsvInputTest.java:128)
```

but not locally

tinwelint committed:
Removed use of flaky Mockito spy

for checking whether or not close() had been called on an anonymous
class, worked locally, but not on build server. Now instead asserts
closed by issuing a call which contractually fails with a specific
exception if closed.

chrisvest commented:
Anonymous inner classes are always final. It should fail for you because of this change:

```
-        return new BufferedCharSeeker( wrap( new StringReader( data ) ), SEEKER_CONFIG );
+        return wrap( new StringReader( data ) );
```

chrisvest commented:
At any rate, less mocking usually means better tests.

tinwelint commented:
Right, although I'm curious to know why it passes locally, but I won't spend more time investigating it. Lesson learned.

tinwelint committed:
Fixes a race in shutdown of DynamicTaskExecutor

where shutdown() might race with processors() and the processor
would forever miss the fact that it had been shutdown.

tinwelint commented:
@chrisvest rebased now

chrisvest commented on community/csv/src/main/java/org/neo4j/csv/reader/CharReadable.java:
```diff
@@ -56,6 +56,13 @@
     SectionedCharBuffer read( SectionedCharBuffer buffer, int from ) throws IOException;
 
+    /**
+     * Reads characters into the given array starting at {@code offset}, reading {@code length} number of bytes.
```
by "bytes" you mean "chars"?

chrisvest commented on community/csv/src/main/java/org/neo4j/csv/reader/Extractors.java:
```diff
@@ -248,6 +248,12 @@ public String toString()
+        @Override
+        public Extractor<T> clone()
+        {
+            throw new UnsupportedOperationException();
```
This should be `CloneNotSupportedException`

chrisvest commented on community/csv/src/main/java/org/neo4j/csv/reader/Extractors.java:
```diff
@@ -313,6 +319,12 @@ public String value()
+        public Extractor<String> clone()
+        {
+            return new StringExtractor( emptyStringsAsNull );
+        }
```
You rely on the clones to have reset their internal mutable state?

chrisvest replied:
Otherwise I think you can rely on `Object.clone()`

tinwelint replied:
Did not know that, will simplify this

chrisvest commented on community/csv/src/main/java/org/neo4j/csv/reader/ProcessingSource.java:
```diff
@@ -0,0 +1,220 @@
+
+    // Buffer reuse, grows on demand up to number of processors concurrently parsing character data
+    private final int maxNumberOfBufferedChunks;
+    private volatile char[][] buffers = new char[3][];
```
The `volatile` indications that this class will face concurrent access, but I don't see any other form of synchronisation going on in here.

chrisvest commented:
First review round done. It's pretty big. Apart from the comments, I'm not spotting anything untoward.

tinwelint committed:
Provides correct CSV line numbers and position when processing in parallel

by rebasing those in the thread iterating over the batches.
The numbers are local in each batch and hence needs to be rebased.

tinwelint commented:
@chrisvest comments addressed and main change set slightly reduced due to removal of cloning code. A couple of added tests as well as a clearer buffer reuse strategy in `ProcessingSource`.
